shared/quantum/codeql/quantum/experimental/Model.qll

```diff
@@ -0,0 +1,2596 @@
+    MD4() or
+    MD5() or
+    MDC2() or
+    POLY1305() or
+    SHA1() or
+    SHA2() or
+    SHA3() or
+    SHAKE() or
+    SM3() or
+    RIPEMD160() or
+    WHIRLPOOL() or
+    OtherHashType()
+
+  /**
+   * A hashing algorithm that transforms variable-length input into a fixed-size hash value.
+   */
+  final class HashAlgorithmNode extends AlgorithmNode, THashAlgorithm {
+    HashAlgorithmInstanceOrValueConsumer instance;
+
+    HashAlgorithmNode() { this = THashAlgorithm(instance) }
+
+    override string getInternalType() { result = "HashAlgorithm" }
+
+    override LocatableElement asElement() { result = instance }
+
+    override string getRawAlgorithmName() { result = instance.asAlg().getRawHashAlgorithmName() }
+
+    final private predicate hashTypeToNameMapping(THashType type, string name) {
+      type = BLAKE2B() and name = "BLAKE2B"
+      or
+      type = BLAKE2S() and name = "BLAKE2S"
+      or
+      type = RIPEMD160() and name = "RIPEMD160"
+      or
+      type = MD2() and name = "MD2"
+      or
+      type = MD4() and name = "MD4"
+      or
+      type = MD5() and name = "MD5"
+      or
+      type = POLY1305() and name = "POLY1305"
+      or
+      type = SHA1() and name = "SHA1"
+      or
+      type = SHA2() and name = "SHA2"
+      or
+      type = SHA3() and name = "SHA3"
+      or
+      type = SHAKE() and name = "SHAKE"
+      or
+      type = SM3() and name = "SM3"
+      or
+      type = WHIRLPOOL() and name = "WHIRLPOOL"
+    }
+
+    /**
+     * Gets the type of this hashing algorithm, e.g., MD5 or SHA.
+     *
+     * When modeling a new hashing algorithm, use this predicate to specify the type of the algorithm.
+     */
+    THashType getHashFamily() { result = instance.asAlg().getHashFamily() }
+
+    override string getAlgorithmName() { this.hashTypeToNameMapping(this.getHashFamily(), result) }
+
+    int getDigestLength() {
+      result = instance.asAlg().getFixedDigestLength() or
+      fixedImplicitDigestLength(instance.asAlg().getHashFamily(), result)
+    }
+
+    final override predicate properties(string key, string value, Location location) {
+      super.properties(key, value, location)
+      or
+      // [KNOWN_OR_UNKNOWN]
+      key = "DigestSize" and
+      if exists(this.getDigestLength())
+      then value = this.getDigestLength().toString() and location = this.getLocation()
+      else (
+        value instanceof UnknownPropertyValue and location instanceof UnknownLocation
+      )
+    }
+  }
+
+  /**
+   * Elliptic curve algorithms
+   */
+  newtype TEllipticCurveType =
+    NIST() or
+    SEC() or
+    NUMS() or
+    PRIME() or
+    BRAINPOOL() or
+    CURVE25519() or
+    CURVE448() or
+    C2() or
+    SM2() or
+    ES() or
+    OtherEllipticCurveType()
+
+  private predicate isBrainpoolCurve(string curveName, int keySize) {
+    // ALL BRAINPOOL CURVES
+    keySize in [160, 192, 224, 256, 320, 384, 512] and
+    (
+      curveName = "BRAINPOOLP" + keySize.toString() + "R1"
```

Comment:
No need for the explicit `.toString()`. The same goes for similar cases in other predicates.